optimism/src/trace.rs

```diff
@@ -0,0 +1,62 @@
+use crate::{lookups::Lookup, E};
+use ark_ff::Field;
+use kimchi_msm::witness::Witness;
+use std::{collections::HashMap, hash::Hash};
+
+pub struct Trace<const N: usize, SELECTOR, F> {
+    /// The domain size of the circuit
+    pub domain_size: usize,
+    /// The witness for a given selector
+    pub witness: HashMap<SELECTOR, Witness<N, Vec<F>>>,
+    /// The vector of constraints for a given selector
+    pub constraints: HashMap<SELECTOR, Vec<E<F>>>,
+    /// The vector of lookups for a given selector
+    pub lookups: HashMap<SELECTOR, Vec<Lookup<E<F>>>>,
+}
+
+impl<const N: usize, SELECTOR: Eq + Hash, F: Field> Trace<N, SELECTOR, F> {
+    /// Returns a boolean indicating whether the witness for the given selector is empty.
+    pub fn witness_is_empty(&self, step: SELECTOR) -> bool {
+        self.witness[&step].cols[0].is_empty()
+    }
+
+    /// Resets the witness after folding
+    pub fn reset(&mut self, step: SELECTOR) {
+        self.witness.insert(
+            step,
+            Witness {
+                cols: Box::new(std::array::from_fn(|_| {
+                    Vec::with_capacity(self.domain_size)
+                })),
+            },
+        );
+    }
+}
+
+pub trait Tracer<const N: usize, SELECTOR, F, Env> {
```

Comment:
Might be obvious, but is `N` the number of columns, for the given selector? If yes, can you add a small comment please?

Reply:
Would it be possible to document each type parameter please? A single line should be enough like:
```
A (F-)tracer builds traces for some program executions.
The type parameter `F` is the type the data points in the trace are encoded into. It can be a field or a native type (u64).
The type `SELECTOR` encodes the information of the kind of information the trace encodes. For instance, if `SELECTOR` is somehow `MIPS_ADD`, the trace could be interpreted as an encoding of the execution of the CPU instruction of a MIPS machine.
The type parameter `N` is defined as the number of columns/"registers" the trace can use per row.
...
```

Reply:
Yes, that's it. Probably given that SELECTOR is verbose, I can also rename N for COLUMNS instead.